src/main/java/com/rcv/Main.java

```diff
@@ -151,97 +166,55 @@ static void executeTabulation(ElectionConfig config) {
   private static List<CastVoteRecord> parseCastVoteRecords(ElectionConfig config) {
+    Logger.tabulationLog(Level.INFO, "Parsing cast vote records...");
+
     // castVoteRecords will contain all cast vote records parsed by the reader
     List<CastVoteRecord> castVoteRecords = new ArrayList<>();
-    // how many sources had fatal problems?
-    int numInvalidSources = 0;
-
-    if (config.rawConfig.cvrFileSources == null || config.rawConfig.cvrFileSources.isEmpty()) {
-      Logger.tabulationLog(Level.SEVERE, "Config doesn't contain any CVR input files.");
-    } else {
-      // At each iteration of the following loop, we add records from another source file.
-      // source: index over config sources
-      for (RawElectionConfig.CVRSource source : config.rawConfig.cvrFileSources) {
-        Logger.tabulationLog(
-            Level.INFO,
-            "Reading CVR file: %s (provider: %s)",
-            source.getFilePath(),
-            source.getProvider());
-
-        // did we encounter a fatal problem for this source?
-        boolean encounteredProblemForThisSource = false;
-
-        if (source.getFilePath() == null || source.getFilePath().isEmpty()) {
-          Logger.tabulationLog(Level.SEVERE, "Invalid source file: missing filePath");
-          encounteredProblemForThisSource = true;
-        }
+    // did we encounter a fatal problem for this source?
+    boolean encounteredSourceProblem = false;
+
+    // At each iteration of the following loop, we add records from another source file.
+    // source: index over config sources
+    for (RawElectionConfig.CVRSource source : config.rawConfig.cvrFileSources) {
+      String logMessage = String.format("Reading CVR file: %s", source.getFilePath());
+      if (source.getProvider() != null && !source.getProvider().isEmpty()) {
+        logMessage += String.format(" (provider: %s)", source.getProvider());
+      }
+      Logger.tabulationLog(Level.INFO, logMessage);
 
-        if (source.getFirstVoteColumnIndex() == null || source.getFirstVoteColumnIndex() < 0) {
+      // the CVRs parsed from this source
+      try {
+        List<CastVoteRecord> cvrs = new CVRReader(config, source).parseCVRFile();
+        if (cvrs.isEmpty()) {
           Logger.tabulationLog(
-              Level.SEVERE,
-              "Invalid source file: missing or invalid firstVoteColumnIndex: %s",
-              source.getFilePath());
-          encounteredProblemForThisSource = true;
+              Level.SEVERE, "Source file contains no CVRs: %s", source.getFilePath());
+          encounteredSourceProblem = true;
         }
-
-        if (config.isTabulateByPrecinctEnabled()
-            && (source.getPrecinctColumnIndex() == null || source.getPrecinctColumnIndex() < 0)) {
+        // add records to the master list
+        castVoteRecords.addAll(cvrs);
+      } catch (SourceWithUnrecognizedCandidatesException exception) {
```

Comment:
This is a cool idea.  It could be used with #60 to flag problems with cvr and candidate configuration if/when we get to that.